### PR TITLE
add delete teams endpoint

### LINKED ISSUE #29: Implement delete endpoint for team

### PR DESCRIPTION
**Pull request by adamszcz**
# What I added (link any issues addressed)
Closes #29 
# How to test it

- Ensure there are teams in the database
- Call ```DELETE https://localhost:5001/api/team/delete```
- Include the following as the body

```
{
    "id": int
}
```

- See that it deletes the record and sets the players to not having a team

### CODE REVIEW

File: CribblyBackend.UnitTests/ControllerTests/TeamControllerTests.cs [hunk 1]
```diff
@@ -122,5 +122,21 @@ public async Task Create_ShouldReturn500_IfError()
             var typedResult = Assert.IsType<ObjectResult>(result);
             Assert.Equal(StatusCodes.Status500InternalServerError, typedResult.StatusCode);
         }
+
+        [Fact]
+        public async Task Delete_ShouldReturnOk_IfNoError()
+        {
+            mockTeamService.Setup(x => x.Delete(It.IsAny<Team>()));
+            var result = await TeamController.Delete(new Team());
+            Assert.IsType<OkResult>(result);
```
Comment (cszczepaniak) [inline CribblyBackend.UnitTests/ControllerTests/TeamControllerTests.cs]: I _think_ there might be a status code that's like "no content" or something that may be more appropriate for delete

File: src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs [hunk 2]
```diff
@@ -63,9 +63,28 @@ public async Task<int> Create(Team team)
             return (await _connection.QueryAsync<int>(@"SELECT LAST_INSERT_ID()")).First();
         }
 
-        public void Delete(Team team)
+        public async Task Delete(Team team)
         {
-            throw new System.NotImplementedException();
+            if (this.GetById(team.Id).Result == null)
+            {
+                throw new System.ArgumentNullException("Team not found, nothing to delete");
```
Comment (cszczepaniak) [inline src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs]: This maybe isn't the most appropriate exception. Maybe we want a custom one?

File: src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs [hunk 3]
```diff
@@ -63,9 +63,28 @@ public async Task<int> Create(Team team)
...
+            };
+
+            await _connection.ExecuteAsync(
+                @"
+                    SET foreign_key_checks = 0;
```
Comment (cszczepaniak) [inline src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs]: Is this a good strategy in production?

File: src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs [hunk 4]
```diff
@@ -63,9 +63,28 @@ public async Task<int> Create(Team team)
...
+
+                    UPDATE cribblydev.players
+                    SET cribblydev.players.TeamId = 0
```
Comment (cszczepaniak) [inline src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs]: Why hardcode the db name here?

Comment (adamszcz) [reply src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs]: Ahhh forgot to delete that out.

File: CribblyBackend.UnitTests/ControllerTests/TeamControllerTests.cs (hunk 1 above)
Comment (adamszcz) [reply CribblyBackend.UnitTests/ControllerTests/TeamControllerTests.cs]: Changed

Comment (adamszcz) [reply CribblyBackend.UnitTests/ControllerTests/TeamControllerTests.cs]: Changed

File: src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs (hunk 3 above)
Comment (adamszcz) [reply src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs]: I have yet to find a better way to do it. I will keep digging

Comment (cszczepaniak) [reply src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs]: Maybe using a transaction would help?

Comment (cszczepaniak) [reply src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs]: Probably the thing you'd have to do is reverse the order -- set the ID to null on the player, then delete the team.

Comment (adamszcz) [reply src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs]: I figured out my issue. I needed to set the TeamId to null instead of 0. 🤦

**Comment by adamszcz** [comment]
Is this bad boy good to go?

File: src/CribblyBackend.DataAccess/Exceptions/TeamNotFoundException.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+            : base("Team not found")
+        {
+            
+        }
+
+        public TeamNotFoundException(string message, Exception inner)
+            : base("Team not found", inner)
+        {
+            
+        }
```
Comment (cszczepaniak) [inline src/CribblyBackend.DataAccess/Exceptions/TeamNotFoundException.cs]: I'd say we probably want one constructor similar to this:
```
public TeamNotFoundException(int id)
    : base($"Team with id {id} not found")
{
}

File: src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs [hunk 6]
```diff
@@ -63,9 +64,24 @@ public async Task<int> Create(Team team)
             return (await _connection.QueryAsync<int>(@"SELECT LAST_INSERT_ID()")).First();
         }
 
-        public void Delete(Team team)
+        public async Task Delete(Team team)
         {
-            throw new System.NotImplementedException();
+            if (this.GetById(team.Id).Result == null)
```
Comment (cszczepaniak) [inline src/CribblyBackend.DataAccess/Repositories/TeamRepository.cs]: 📖 I'd probably drop the `this.`

File: src/CribblyBackend/Controllers/TeamController.cs [hunk 7]
```diff
@@ -97,5 +99,33 @@ public async Task<IActionResult> GetByTeamId(int id)
...
             return NotFound();
         }
+
+        /// <summary>
+        /// Delete permanently removes the team with the specified id.
+        /// </summary>
+        /// <param name="id">The Team object that will be deleted from the database</param>
+        /// <returns></returns>
+        [HttpDelete("{id}")]
+        public async Task<IActionResult> Delete([FromBody] Team team)
```
Comment (cszczepaniak) [inline src/CribblyBackend/Controllers/TeamController.cs]: Let's choose _either_ a URL param or something in the body. I think for deletes, it's more common to put the ID in the URL.

Comment (adamszcz) [reply src/CribblyBackend/Controllers/TeamController.cs]: Fixed?

**Comment by adamszcz** [comment]
I think this one should be all set!

File: src/CribblyBackend.DataAccess/Teams/Repositories/TeamRepository.cs [hunk 8]
```diff
@@ -89,10 +85,17 @@ UPDATE players
                     SET players.TeamId = null
                     WHERE TeamId = @Id;
 
+                    DELETE FROM scores
+                    WHERE TeamId = @Id;
+
+                    UPDATE games
+                    SET games.WinnerId = null
+                    WHERE WinnerId = @Id;
+
```
Comment (cszczepaniak) [inline src/CribblyBackend.DataAccess/Teams/Repositories/TeamRepository.cs:94]: Would it make sense to do all of these operations in a transaction?

**Review by cszczepaniak — APPROVED** [review]
We could use a transaction when doing several operations to delete a team, but I'm not going to hold this one up for that

**Pull request merged**